test/e2e/load_balancer_test.go

```diff
@@ -0,0 +1,151 @@
+		jig.SanityCheckService(tcpService, v1.ServiceTypeLoadBalancer)
+		tcpNodePortOld := tcpNodePort
+		tcpNodePort = int(tcpService.Spec.Ports[0].NodePort)
+		if tcpNodePort == tcpNodePortOld {
+			framework.Failf("TCP Spec.Ports[0].NodePort (%d) did not change", tcpNodePort)
+		}
+		if framework.GetIngressPoint(&tcpService.Status.LoadBalancer.Ingress[0]) != tcpIngressIP {
+			framework.Failf("TCP Status.LoadBalancer.Ingress changed (%s -> %s) when not expected", tcpIngressIP, framework.GetIngressPoint(&tcpService.Status.LoadBalancer.Ingress[0]))
+		}
+		framework.Logf("TCP node port: %d", tcpNodePort)
+
+		By("hitting the TCP service's new NodePort")
+		jig.TestReachableHTTP(nodeIP, tcpNodePort, framework.KubeProxyLagTimeout)
+
+		By("checking the old TCP NodePort is closed")
+		jig.TestNotReachableHTTP(nodeIP, tcpNodePortOld, framework.KubeProxyLagTimeout)
+
+		By("hitting the TCP service's LoadBalancer")
+		jig.TestReachableHTTP(tcpIngressIP, svcPort, loadBalancerLagTimeout)
+
+		// Change the services' main ports.
+
+		By("changing the TCP service's port")
+		tcpService = jig.UpdateServiceOrFail(ns, tcpService.Name, func(s *v1.Service) {
+			s.Spec.Ports[0].Port++
+		})
+		jig.SanityCheckService(tcpService, v1.ServiceTypeLoadBalancer)
+		svcPortOld := svcPort
+		svcPort = int(tcpService.Spec.Ports[0].Port)
+		if svcPort == svcPortOld {
+			framework.Failf("TCP Spec.Ports[0].Port (%d) did not change", svcPort)
+		}
+		if int(tcpService.Spec.Ports[0].NodePort) != tcpNodePort {
+			framework.Failf("TCP Spec.Ports[0].NodePort (%d) changed", tcpService.Spec.Ports[0].NodePort)
+		}
+		if framework.GetIngressPoint(&tcpService.Status.LoadBalancer.Ingress[0]) != tcpIngressIP {
+			framework.Failf("TCP Status.LoadBalancer.Ingress changed (%s -> %s) when not expected", tcpIngressIP, framework.GetIngressPoint(&tcpService.Status.LoadBalancer.Ingress[0]))
+		}
+
+		framework.Logf("service port (TCP): %d", svcPort)
+
+		By("hitting the TCP service's NodePort")
+		jig.TestReachableHTTP(nodeIP, tcpNodePort, framework.KubeProxyLagTimeout)
+
+		By("hitting the TCP service's LoadBalancer")
+		jig.TestReachableHTTP(tcpIngressIP, svcPort, loadBalancerCreateTimeout) // this may actually recreate the LB
+
+		// Change the services back to ClusterIP.
+
+		By("changing TCP service back to type=ClusterIP")
+		tcpService = jig.UpdateServiceOrFail(ns, tcpService.Name, func(s *v1.Service) {
+			s.Spec.Type = v1.ServiceTypeClusterIP
+			s.Spec.Ports[0].NodePort = 0
+		})
+		// Wait for the load balancer to be destroyed asynchronously
+		tcpService = jig.WaitForLoadBalancerDestroyOrFail(ns, tcpService.Name, tcpIngressIP, svcPort, loadBalancerCreateTimeout)
+		jig.SanityCheckService(tcpService, v1.ServiceTypeClusterIP)
+
+		By("checking the TCP NodePort is closed")
+		jig.TestNotReachableHTTP(nodeIP, tcpNodePort, framework.KubeProxyLagTimeout)
+
+		By("checking the TCP LoadBalancer is closed")
+		jig.TestNotReachableHTTP(tcpIngressIP, svcPort, loadBalancerLagTimeout)
+	})
+})
```

Comment:
We should create a ticket to track testing preserving the source ip once OCI supports tcp proxy protocol.

Reply:
https://github.com/oracle/oci-cloud-controller-manager/issues/139